pkg/spec/util.go

```diff
@@ -0,0 +1,215 @@
+package spec
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	kedgeSpecPackagePath = "github.com/kedgeproject/kedge/pkg/spec"
+)
+
+// This function takes a map as an input which stores the JSON tag as key and
+// the names of the structs holding that tag as the value. The passed in
+// blacklisted tags are ignored and are not populated at all.
+func getUnmarshalJSONTagsMap(inputMap map[string][]string, inputStruct interface{}, blacklistedTags []string) error {
+
+	// Checking if input is pointer to struct
+	if err := checkTypePointerToStruct(inputStruct); err != nil {
+		return errors.Wrap(err, "Input parameter type mismatch")
+	}
+
+	val := reflect.ValueOf(inputStruct).Elem()
+
+StructFieldsLoop:
+	for i := 0; i < val.NumField(); i++ {
+		field := val.Type().Field(i)
+
+		// We do not need to store JSON tags like "inline" or "omitempty",
+		// instead we only need JSON tag which is used for unmarshalling the
+		// YAML. It's safe to assume that, that JSON tag is the one that
+		// is specified first.
+		// Let's split the JSON tag on ",", and store the first element of the
+		// resulting array.
+		unmarshalJSONTag := strings.SplitN(field.Tag.Get("json"), ",", 2)[0]
+
+		if unmarshalJSONTag == "" {
+			// It does not make any sense to store anything for JSON tags with no
+			// first element, e.g. `json:",inline"`
+			continue
+		}
+
+		// Populating only the non blacklisted tags
+		for _, blacklistedTag := range blacklistedTags {
+			if unmarshalJSONTag == blacklistedTag {
+				continue StructFieldsLoop
+			}
+		}
```

Comment:
if `blacklistedTags` this loop for searching won't be needed and the search will become `O(1)`